File: module_1/python3/StockAnalysis.py
import re

import pandas as pd


class StockAnalysis:
    dateformat = '%d-%b-%Y'

    def __init__( self, filename: str ):
        ### 1.1 Import the csv file of the stock of your choosing using 'pd.read_csv()' function into a dataframe.
        self.filename = filename
        self.data = pd.read_csv(self.filename, parse_dates=['Date'] )        # parse_dates=['Date'] cast to Timestamp
        self.data = self.data.rename(lambda x: re.sub(r'[^\w\s]+', '', x).replace(r' ', '_'), axis=1)  # Rename columns without spaces
        self.data = self.filter_not_eq(self.data)

        ### 1.3 Change the date column from 'object' type to 'datetime64(ns)' | WORKAROUND: cast to pd.Timestamp instead
        # self.data['Date'] = pd.to_datetime(self.data['Date'])              # Cast to Timestamp without parse_dates=["Date"]
        # self.data['Date'] = self.data['Date'].map(lambda date: np.datetime64(date.value, 'ns')) # BUG: Cast to datetime64 returns Timestamp
        # print( type(self.data['Date'][0]), self.data['Date'][0] )                               # BUG: Cast to datetime64 returns Timestamp

        ### 1.4 Hint : Create a new dataframe column ‘Month’ + 'Year'
        self.data['Date_Month'] = self.data['Date'].map(lambda date: date.month)
        self.data['Date_Year']  = self.data['Date'].map(lambda date: date.year)


    def print(self):
        print(self.__class__.__name__)
        print(self.filename)
        print('\nself.data.head()\n',     self.data.head())
        print('\nself.data.describe()\n', self.data.describe())
        print('\nself.stats_90_day_close_price()\n', self.stats_90_day_close_price())
        print('\nself.stats_vwap()\n', self.stats_vwap_by_month())


    @staticmethod
    def filter_not_eq( data: pd.DataFrame ) -> pd.DataFrame:
        """  1.1 remove all the rows where 'Series' column is NOT 'EQ' """
        return data[data.Series != 'EQ']


    @staticmethod
    def filter_days( data: pd.DataFrame, days=90 ) -> pd.DataFrame:
        date_end:    pd.Timestamp = data['Date'].max()
        date_cutoff: pd.Timestamp = date_end - pd.to_timedelta(90, unit='d')
        return data[data.Date > str(date_cutoff)]


    def stats_90_day_close_price(self) -> dict:
        """ 1.2 Calculate the maximum, minimum and mean price for the last 90 days. (price=Closing Price unless stated otherwise) """
        data_90_days = self.filter_days(self.data, 90)
        stats = {
            "min":  data_90_days['Close_Price'].min(),
            "max":  data_90_days['Close_Price'].max(),
            "mean": data_90_days['Close_Price'].mean(),
        }
        stats = { k: round(v,2) for k,v in stats.items() }  # round to 2dp
        return stats


    @staticmethod
    def vwap(group_df: pd.DataFrame) -> float:
        vwap = (group_df.Close_Price + group_df.Total_Traded_Quantity).sum() \
               / group_df.Total_Traded_Quantity.sum()
        return vwap


    def stats_vwap_by_month(self) -> pd.DataFrame:
        """
        1.4 Calculate the monthwise VWAP = sum(price*volume)/sum(volume)
        :return:
        """
        vwap = {}
        groupByMonth = self.data.groupby(['Date_Year', 'Date_Month'])
        for key, group_df in groupByMonth:
            vwap[key] = {
                "Year":  int(key[0]),
                "Month": int(key[1]),
                "VWAP":  self.vwap(group_df)
            }
        vwap_df = pd.DataFrame(vwap).transpose().reindex(["Year", "Month", "VWAP"], axis=1)
        vwap_df[["Year", "Month"]] = vwap_df[["Year", "Month"]].apply(pd.to_numeric, downcast='integer')
        return vwap_df

File: scala/src/main/scala/module_1/StockAnalysis.scala
package module_1

import java.time.LocalDate

import pprint.pprintln

import scala.Numeric._
import scala.collection.immutable.{ListMap, Map}
//import scala.util.chainingOps._  // Requires scala v2.13 - incompatible with kantan

object StockAnalysis extends App {
  val input_csv_filename:   String = if (args.length >= 1) args(0) else "../data/stocks/Mid_Cap/MUTHOOTFIN.csv"
  val output_csv_filename:  String = if (args.length >= 2) args(1) else "../data/output/week2-scala.csv"
  val output_json_filename: String = if (args.length >= 3) args(2) else "../data/output/week2-scala.json"

  var stockPrices: List[StockPrice] = StockPriceCSV.read(input_csv_filename)

  stockPrices = filter_not_eq(stockPrices)
  StockAnalysis.print(stockPrices)
  StockPriceCSV.write_csv(output_csv_filename,    stockPrices)
  StockPriceCSV.write_json(output_json_filename, stats(stockPrices))

  def print( stockPrices: List[StockPrice] ) {
    println(this.getClass.getSimpleName)
    println(input_csv_filename)
    println("\nstockPrices.take(5)");                 pprintln(stockPrices.take(5))

    pprintln( stats(stockPrices) )
  }

  def stats( stockPrices: List[StockPrice] ): Map[String, Map[String, Double]]  = {
    Map(
      "stats_90_day_close_price"     -> stats_90_day_close_price(stockPrices),
      "stats_vwap_by_month"          -> stats_vwap_by_month(stockPrices),
      "stats_average_price"          -> stats_average_price(stockPrices),
      "stats_profit_loss_percentage" -> stats_profit_loss_percentage(stockPrices),
      // "stats_quantity_trend"         -> stats_quantity_trend(stockPrices)
    )
  }


  def filter_not_eq( stockPrices: List[StockPrice]): List[StockPrice] = {
    stockPrices.filter( _.Series != "EQ" )
  }

  def filter_days( stockPrices: List[StockPrice], days: Int ): List[StockPrice] = {
    // DOCS: https://stackoverflow.com/questions/38059191/how-make-implicit-ordered-on-java-time-localdate
    implicit val localDateOrdering: Ordering[LocalDate] = _ compareTo _

    val date_end:    LocalDate = stockPrices.map(_.Date).max
    val date_cutoff: LocalDate = date_end.minusDays(days)

    stockPrices.filter(_.Date.compareTo(date_cutoff) > 0)
  }

  def vwap( stockPrices: List[StockPrice] ): Double = {
    val total = stockPrices.map(stockPrice => stockPrice.Close_Price * stockPrice.Total_Traded_Quantity).sum
    val price = stockPrices.map(stockPrice => stockPrice.Total_Traded_Quantity).sum
    total / price
  }

  def stats_vwap_by_month( stockPrices: List[StockPrice] ): ListMap[String,Double] = {
    val vwap_by_month = stockPrices
      .groupBy(stockPrice => f"${stockPrice.Date.getYear}-${stockPrice.Date.getMonthValue}%02d")
      .mapValues(group => vwap(group))

    ListMap( vwap_by_month.toSeq.sortBy(_._1):_* )  // months need to be zero padded to sort correctly
  }

  // 1.5 Write a function to calculate the average price over the last N days of the stock price data where N is a user defined parameter.
  def average_price(stockPrices: List[StockPrice], days: Int): Double = {
    val prices = filter_days(stockPrices, days).map(_.Close_Price)

    prices.length match {
      case 0 => 0
      case 1 => 0
      case _ => grizzled.math.stats.mean( prices.head, prices.tail:_* )  // Weird syntax: https://github.com/bmc/grizzled-scala/blob/master/src/test/scala/grizzled/math/StatsSpec.scala
    }
  }

  // 1.5 Write a second function to calculate the profit/loss percentage over the last N days
  def profit_loss_percentage(stockPrices: List[StockPrice], days: Int): Double = {
    val prices = filter_days(stockPrices, days).map(_.Close_Price)

    prices.length match {
      case 0 => 0
      case 1 => 0
      case _ => (prices.last - prices.head) / prices.head * 100
    }
  }

  // 1.2 Calculate the maximum, minimum and mean price for the last 90 days. (price=Closing Price unless stated otherwise)
  def stats_90_day_close_price(stockPrices: List[StockPrice]): ListMap[String, Double] = {
    val prices = filter_days(this.stockPrices, 90).map(_.Close_Price)

    ListMap(
      "min"  -> prices.min,
      "max"  -> prices.max,
      "mean" -> grizzled.math.stats.mean( prices.head, prices.tail:_* )
    )
  }

  /**
    * 1.5 Calculate the average price AND the profit/loss percentages over the course of
    * last - 1 week, 2 weeks, 1 month, 3 months, 6 months and 1 year.
    */
  def stats_average_price( stockPrices: List[StockPrice] ): ListMap[String, Double] = {
    ListMap(
      "1 week"   -> average_price(stockPrices, 7 * 1),
      "2 weeks"  -> average_price(stockPrices, 7 * 2),
      "1 month"  -> average_price(stockPrices, 365 / 12 * 1),
      "2 months" -> average_price(stockPrices, 365 / 12 * 2),
      "6 months" -> average_price(stockPrices, 365 / 12 * 6),
      "1 year"   -> average_price(stockPrices, 365),
    )
  }

  /**
    * 1.5 Calculate the average price AND the profit/loss percentages over the course of
    * last - 1 week, 2 weeks, 1 month, 3 months, 6 months and 1 year.
    */
  def stats_profit_loss_percentage(stockPrices: List[StockPrice] ): ListMap[String, Double] = {
    ListMap(
      "1 week"   -> profit_loss_percentage(stockPrices, 7 * 1),
      "2 weeks"  -> profit_loss_percentage(stockPrices, 7 * 2),
      "1 month"  -> profit_loss_percentage(stockPrices, 365 / 12 * 1),
      "2 months" -> profit_loss_percentage(stockPrices, 365 / 12 * 2),
      "6 months" -> profit_loss_percentage(stockPrices, 365 / 12 * 6),
      "1 year"   -> profit_loss_percentage(stockPrices, 365),
    )
  }

//  // 1.8: Find the average and median values of the column 'Total Traded Quantity' for each of the types of 'Trend'.
//  def stats_quantity_trend( stockPrices: List[StockPrice] ) {
//    stockPrices
//        .groupBy(_.Trend)
//
//    const trends = _(this.data)
//      .groupBy('Trend')
//    .toPairs().sortBy(0).fromPairs() // Sort by keys
//      .mapValues((group) => _.map(group, 'Total_Traded_Quantity'))
//    .mapValues((values) => ({
//      "mean": ss.mean(values),
//      "median": ss.median(values),
//    }))
//      .value();
//    return trends;
//  }
}
